src/Concretize.hs

```diff
@@ -36,347 +59,398 @@ data Level = Toplevel | Inside
-                lambdaPath = SymPath spath ("_Lambda_" ++ lambdaToCName name (envFunctionNestingLevel envWithArgs) ++ "_" ++ show (infoIdentifier ii) ++ "_env")
-                lambdaNameSymbol = XObj (Sym lambdaPath Symbol) (Just dummyInfo) Nothing
-                -- Anonymous functions bound to a let name might call themselves. These recursive instances will have already been qualified as LookupRecursive symbols.
-                -- Rename the recursive calls according to the generated lambda name so that we can call these correctly from C.
-                renameRecursives (XObj (Sym _ LookupRecursive) si st) = (XObj (Sym lambdaPath LookupRecursive) si st)
-                renameRecursives x = x
-                recBody = walk renameRecursives okBody
-                environmentTypeName = pathToC lambdaPath ++ "_ty"
-                tyPath = (SymPath [] environmentTypeName)
-                extendedArgs =
-                  if null capturedVars
-                    then args
-                    else -- If the lambda captures anything it need an extra arg for its env:
-
-                      XObj
-                        ( Arr
-                            ( XObj
-                                (Sym (SymPath [] "_env") Symbol)
-                                (Just dummyInfo)
-                                (Just (PointerTy (StructTy (ConcreteNameTy tyPath) []))) :
-                              argsArr
-                            )
-                        )
-                        ai
-                        at
-                lambdaCallback = XObj (Lst [XObj (Defn (Just (Set.fromList capturedVars))) (Just dummyInfo) Nothing, lambdaNameSymbol, extendedArgs, recBody]) i t
-                -- The lambda will also carry with it a special made struct containing the variables it captures
-                -- (if it captures at least one variable)
-                structMemberPairs =
-                  concatMap
-                    ( \(XObj (Sym path _) _ (Just symTy)) ->
-                        [XObj (Sym path Symbol) Nothing Nothing, reify symTy]
-                    )
-                    capturedVars
-                environmentStructTy = StructTy (ConcreteNameTy tyPath) []
-                environmentStruct =
-                  XObj
-                    ( Lst
-                        [ XObj (Deftype environmentStructTy) Nothing Nothing,
-                          XObj (Sym tyPath Symbol) Nothing Nothing,
-                          XObj (Arr structMemberPairs) Nothing Nothing
-                        ]
+    go b = pure [defn, name, args, b]
+visitDefn _ _ Inside _ _ x@(ListPat (DefnPat _ _ _ _)) =
+  pure (Left (DefinitionsMustBeAtToplevel x))
+visitDefn _ _ _ _ _ x = pure (Left (CannotConcretize x))
+
+-- | Concretely type a program entry point. Can only return Int or Unit.
+visitMain :: Visitor
+visitMain vs _ Toplevel tenv env (ListPat (DefnPat defn name@(SymPat (SymPath [] "main") _) args@(ArrPat []) body)) =
```

Comment:
The `vs` variable name is very cryptic to me, what does it stand for?

Reply:
it's just the visitedDefinitions -- it's not used in most cases but needs to be threaded through for the one case in which it matters (mkLambda)